### ISSUE TITLE
<PHP> is rendered into a <p> Element

### ISSUE THREAD

**Issue by fiwswe** [fiwswe/dokuwiki-plugin-embeddedphp#1]
When using the block element `<PHP>` syntax the resulting HTML is:
```
<p>
<div class="embeddedphp">output generated by the embedded PHP code</div>
</p>
```
Some browsers will interpret this as:
```
<p></p>
<div class="embeddedphp">output generated by the embedded PHP code</div>
<p></p>
```

Note: The code for the `<PHP>` syntax already uses PType `block` (unless there is a bug somewhere).

**Comment by saggi-dw** [comment]
The renderer is always called. You should only react on 'unmatched'.
```
    public function render($mode, Doku_Renderer $renderer, $data)
    {
        if ($mode === 'xhtml') {
            if (is_array($data) && (count($data) > 1)  && ($data[0] == DOKU_LEXER_UNMATCHED)) {
                $this->php($data[1], $renderer, $this->getwrapper());
                return true;
            }
            return true;
        }
        return false;
    }

```
You could simplify the handle:
```
    public function handle($match, $state, $pos, Doku_Handler $handler)
    {
        // If we are parsing a submitted comment. Executing embedded PHP in comments is
        // not a good idea!
        if (isset($_REQUEST['comment'])) {
            return false;
        }
        return [$state, $match];
    }
```
And remove the extra renderer from the phpblock when determining the wrapper extra.

```
    public function  getwrapper() {
        if ($this->getPType() === 'block') {
            return 'pre';
        }
        return 'code';
    }
```
These are just ideas, you know I'm still learning....

**Comment by fiwswe** [comment]
> The renderer is always called. You should only react on 'unmatched'.

You are right. I'll test this and see if it helps. OTOH in my `handle()` method I return `false` for anything other than `DOKU_LEXER_UNMATCHED`. My thinking was that it would prevent `render()` being called for any other state. But I freely admit I don't fully grok the flow of information yet.


> You could simplify the handle:

Not sure about this one. It seems to do the opposite of your first suggestion.

> And remove the extra renderer from the phpblock when determining the wrapper extra.

Yes there is unnecessary code duplication there. Actually I initially tried to use class variables for things like the tag in my first tries but ran into problems and switched to using accessor methods. I probably did something wrong as many other plugins do use class variables, e.g.:
```php
class syntax_plugin_embeddedphp_phpinline extends \dokuwiki\Extension\SyntaxPlugin
{
    protected $tag = 'php';
    protected $wrapper = 'pre';
    …
}
```
and overrides:
```php
class syntax_plugin_embeddedphp_phpblock extends syntax_plugin_embeddedphp_phpinline
{
    protected $tag = 'PHP';  // Overrides base class
    protected $wrapper = 'code';  // Overrides base class
    …
}
```


> These are just ideas, you know I'm still learning....

Thanks! I have some experiments on my todo-list ;-)

I think one of my issues was that I am hosting the wiki on a remote machine which limits the tools I can use for debugging. So it was really hard to figure out at which point which data was actually being used and what code was actually executed. I essentially sprinkled `\dokuwiki\Logger::debug()` statements into my code and then looked at the debug log. To say this is not ideal is like saying the Atlantic Ocean is a puddle of water ;-)

**Comment by saggi-dw** [comment]
> I think one of my issues was that I am hosting the wiki on a remote machine which limits the tools I can use for debugging. So it was really hard to figure out at which point which data was actually being used and what code was actually executed. I essentially sprinkled \dokuwiki\Logger::debug() statements into my code and then looked at the debug log. To say this is not ideal is like saying the Atlantic Ocean is a puddle of water ;-)

I know this, so many `var_dump($var);`

> Not sure about this one. It seems to do the opposite of your first suggestion.

Actually, I'm not sure either. As I understand it, the return from handle($data) is always passed to render.
So the evaluation can also be done in render, it doesn't have to be twice.
But I could be wrong...
Then even a simple 
```php
if (!empty($data) && ($data[0] == DOKU_LEXER_UNMATCHED)) {
    $this->php($data[1], $renderer, $this->getwrapper());
}
```
should be enough.

**Comment by fiwswe** [comment]
See also: [Syntax plugin: Why is block output surrounded by empty paragraphs?](https://forum.dokuwiki.org/d/23317-syntax-plugin-why-is-block-output-surrounded-by-empty-paragraphs)
